### PR TITLE
#604 Bitmap scanner should clear status to enable re-entry

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

When there are multiple tasks use the same scanner, it could lead to core dump.

## How was this patch tested?

add a unit test.

### CODE REVIEW

**Comment by adrian-wang** [comment]
This should be port to branch-0.3...

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapIndexQuerySuite.scala [hunk 1]
```diff
@@ -105,6 +107,26 @@ class OapIndexQuerySuite extends QueryTest with SharedOapContext with BeforeAndA
...
+    val schema =
+      StructType(
+        StructField("a", IntegerType) ::
+          StructField("b", StringType) :: Nil)
+    val df = spark.createDataFrame(rowRDD, schema)
+    df.createOrReplaceTempView("t")
+
+    sql("insert overwrite table oap_test_1 select * from t")
+    sql("create oindex bmidx1 on oap_test_1 (a) using bitmap")
+    sql("create oindex bmidx2 on oap_test_1 (b) using bitmap")
```
Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapIndexQuerySuite.scala]: may not need to create an index bmidx2

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 2]
```diff
@@ -50,17 +50,6 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
 
   private var bmUniqueKeyListTotalSize: Int = _
   private var bmUniqueKeyListCount: Int = _
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Maybe We can get value from bmFooterCache and remove bmUniqueKeyListTotalSize& bmUniqueKeyListCount also, but I'am not sure bmFooterCache is usable in getDesiredBitmapArray method

**Comment by LuciferYang** [comment]
shall we merge this to branch-0.3 also?

**Comment by LuciferYang** [comment]
I think the master and branch-0.3 all need this pr

**Comment by adrian-wang** [comment]
I'll merge this to unblock enabling work on production environment, you can still comment and let's fix it in follow-up pr!

**Comment by adrian-wang** [comment]
merged to master and branch-0.3

**Comment by carsonwang** [comment]
Thanks @adrian-wang for the quick fix. Actually a single partition of the FileScanRDD can contain several files. So a single task may read multiple files using the same BitmapScanner, which caused the issue.

**Comment by LuciferYang** [comment]
Thanks @carsonwang explain this issue & Thanks @adrian-wang fix it~ Already pass the test  in our QA environment~

**Pull request merged**